pkg/castemplate/v1alpha1/kubernetes_test.go

```diff
@@ -0,0 +1,213 @@
+limitations under the License.
+*/
+
+package v1alpha1
+
+import (
+	"reflect"
+	"testing"
+
+	"github.com/pkg/errors"
+
+	apis "github.com/openebs/maya/pkg/apis/openebs.io/v1alpha1"
+	clientset "github.com/openebs/maya/pkg/client/generated/clientset/internalclientset"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+func fakeGetClientset() (cs *clientset.Clientset, err error) {
+	return &clientset.Clientset{}, nil
+}
+
+func fakeGetfn(cs *clientset.Clientset, name string,
+	opts metav1.GetOptions) (*apis.CASTemplate, error) {
+	return &apis.CASTemplate{}, nil
+}
+
+func fakeGetErrfn(cs *clientset.Clientset, name string,
+	opts metav1.GetOptions) (*apis.CASTemplate, error) {
+	return &apis.CASTemplate{}, errors.New("some error")
+}
+
+func fakeSetClientset(k *Kubeclient) {
+	k.clientset = &clientset.Clientset{}
+}
+
+func fakeSetNilClientset(k *Kubeclient) {
+	k.clientset = nil
+}
+
+func fakeGetNilErrClientSet() (clientset *clientset.Clientset, err error) {
+	return nil, nil
+}
+
+func fakeGetErrClientSet() (clientset *clientset.Clientset, err error) {
+	return nil, errors.New("Some error")
+}
+
+func fakeClientSet(k *Kubeclient) {}
+
+func TestWithDefaults(t *testing.T) {
+	tests := map[string]struct {
+		getFn              getFunc
+		getClientsetFn     getClientsetFunc
+		expectGet          bool
+		expectGetClientset bool
+	}{
+		// The current implementation of WithDefaults method can be
+		// tested using these two combinations only.
+		"When mockclient is empty": {nil, nil, false, false},
+		"When mockclient contains all of them": {fakeGetfn,
+			fakeGetClientset, false, false},
+	}
+
+	for name, mock := range tests {
+		t.Run(name, func(t *testing.T) {
+			fc := &Kubeclient{}
+			fc.get = mock.getFn
+			fc.getClientset = mock.getClientsetFn
+
+			fc.withDefaults()
+			get := (fc.get == nil)
+			if get != mock.expectGet {
+				t.Fatalf(`test %s failed: expected non-nil fc.get
+but got %v`, name, fc.get)
+			}
+			getClientset := (fc.getClientset == nil)
+			if getClientset != mock.expectGetClientset {
+				t.Fatalf(`test %s failed: expected non-nil fc.getClientset
+but got %v`, name, fc.getClientset)
+			}
+		})
+	}
+}
+func TestWithClientset(t *testing.T) {
+	tests := map[string]struct {
+		clientSet    *clientset.Clientset
+		isKubeClient bool
+	}{
+		"Clientset is empty":     {nil, false},
+		"Clientset is not empty": {&clientset.Clientset{}, true},
+	}
+
+	for name, mock := range tests {
+		t.Run(name, func(t *testing.T) {
+			h := WithClientset(mock.clientSet)
+			fake := &Kubeclient{}
+			h(fake)
+			if mock.isKubeClient && fake.clientset == nil {
+				t.Fatalf(`test %s failed, expected non-nil fake.clientset
+but got %v`, name, fake.clientset)
+			}
+			if !mock.isKubeClient && fake.clientset != nil {
+				t.Fatalf(`test %s failed, expected nil fake.clientset
+but got %v`, name, fake.clientset)
+			}
+		})
+	}
+}
+func TestKubeClientWithClientset(t *testing.T) {
+	tests := map[string]struct {
+		expectClientSet bool
```

Comment:
Can we reorder the fields just to make it similar to other UT in this repository

Reply:
done